presto-parser/src/main/java/com/facebook/presto/sql/parser/AstBuilder.java

```diff
@@ -1463,6 +1468,76 @@ protected Node aggregateResult(Node aggregate, Node nextResult)
         throw new UnsupportedOperationException("not yet implemented");
     }
 
+    private static String decodeString(SqlBaseParser.UnicodeStringLiteralContext context)
+    {
+        char escape;
+        if (context.UESCAPE() != null) {
+            String escapeString = unquote(context.STRING().getText());
+            check(!escapeString.isEmpty(), "Empty Unicode escape character", context);
+            check(escapeString.length() == 1, "Invalid Unicode escape character: " + escapeString, context);
+            escape = escapeString.charAt(0);
+            check(isValidEscape(escape), "Invalid Unicode escape character: " + escapeString, context);
+        }
+        else {
+            escape = '\\';
+        }
+
+        String rawContent = unquote(context.UNICODE_STRING().getText().substring(2));
+        StringBuilder builder = new StringBuilder();
+        int beginCodeIndex = 0;
+        int charactersNeeded = 0;
+        for (int i = 0; i < rawContent.length(); i++) {
+            if (rawContent.charAt(i) == escape) {
+                check(charactersNeeded == 0, "Incomplete escaped character: " + rawContent.substring(beginCodeIndex, i + 1), context);
+                check(i < rawContent.length() - 1, "Unexpected end of escaped string", context);
+                if (rawContent.charAt(i + 1) == escape) {
+                    builder.append(escape);
+                    i++;
+                }
+                else if (rawContent.charAt(i + 1) == '+') {
+                    charactersNeeded = 6;
+                    i++;
+                }
+                else {
+                    charactersNeeded = 4;
+                }
+                beginCodeIndex = i + 1;
+                continue;
+            }
+
+            if (charactersNeeded != 0) {
+                check(isHexDigit(rawContent.charAt(i)), "Invalid hexadecimal digit: " + rawContent.charAt(i), context);
+                charactersNeeded--;
+            }
+
+            if (charactersNeeded == 0) {
+                if (beginCodeIndex == i) {
+                    // A non-escaped character
+                    builder.append(rawContent.charAt(i));
```

Comment:
I would move this before the `if (charactersNeeded != 0)` check. Otherwise, you will have to manage `beginCodeIndex` when you are not in an escape, which is error-prone.